module/zfs/dmu_send.c

```diff
@@ -3262,12 +3276,18 @@ dmu_recv_end_sync(void *arg, dmu_tx_t *tx)
 	}
 	drc->drc_newsnapobj = dsl_dataset_phys(drc->drc_ds)->ds_prev_snap_obj;
 	zvol_create_minors(dp->dp_spa, drc->drc_tofs, B_TRUE);
+
 	/*
 	 * Release the hold from dmu_recv_begin.  This must be done before
-	 * we return to open context, so that when we free the dataset's dnode,
-	 * we can evict its bonus buffer.
+	 * we return to open context, so that when we free the dataset's dnode
+	 * we can evict its bonus buffer. Since the dataset may be destroyed
+	 * at this point (and therefore won't have a valid pointer to the spa)
+	 * we release the key mapping manually here while we do have a valid
+	 * pointer.
 	 */
-	dsl_dataset_disown(drc->drc_ds, dmu_recv_tag);
+	(void) spa_keystore_remove_mapping(dmu_tx_pool(tx)->dp_spa,
+	    drc->drc_ds, drc->drc_ds);
```

Comment:
So that it releases the key mapping hold it created in `dmu_recv_begin()`. Loading and unloading is for wrapping keys, technically.